src/Language/PureScript/TypeChecker/Kinds.hs

```diff
@@ -889,7 +889,9 @@ checkInstanceDeclaration moduleName (ann, constraints, clsName, args) = do
     ty' <- checkKind ty E.kindConstraint
     constraints' <- for constraints checkConstraint
     allTy <- apply $ foldr srcConstrainedType ty' constraints'
-    allUnknowns <- unknownsWithKinds . IS.toList $ unknowns allTy
+    subst <- gets checkSubstitution
+    let uns = substUnsolved subst
+    allUnknowns <- unknownsWithKinds . IS.toList $ unknowns allTy <> IS.filter (`M.member` uns) (foldMap (unknowns . substituteType subst . snd) freeVarsDict)
```

Comment:
`unknownsWithKinds` walks through the kinds of unknowns as well. The way this is _supposed_ to work is:
* `unknowns` gives you all the unknowns that are actually present in the type.
* `unknownsWithKinds` recursively walks the kinds of unknowns in the environment to find additional unknowns which need to be generalized.

So I wouldn't think this would be necessary, unless `unknownsWithKinds` isn't behaving as expected.

Reply:
See this line which dereferences the unknown's kind and `apply`s it https://github.com/purescript/purescript/blob/6015fe970f9f0fa0a3ad948335e410d5e8f5ced2/src/Language/PureScript/TypeChecker/Kinds.hs#L153

Reply:
Trouble with that is, `unknowns allTy` is empty in this case, so the recursion in `unknownsWithKinds` has nothing to start with.

`allTy` contains free type variables but no unknowns; the unknowns have been placed into the environment as the kinds of the free type variables, but `unknowns` doesn't use the environment; it just looks for unknowns that are used directly in the type.

So we need some step that gets the kinds of the free type variables. We also can't just pass those kinds to `unknowns` and then `unknownsWithKinds`, because `lookupUnsolved` clearly expects its input to still appear in the `substUnsolved` map, and by this point in the checking, those unknowns *have* been solved (but not fully; they've been substituted with different unknowns).

Any thoughts on how the above should happen?

Reply:
> Trouble with that is, unknowns allTy is empty in this case, so the recursion in unknownsWithKinds has nothing to start with.

Ah! Yeah, of course. This is what I was missing :) With that in mind, my thought would be that using `unknowns allTy` is just wrong. Would there ever be any unknowns in allTy (that weren't the result of the unknown kinds of the free type variables)? If that were the case, I might do:
* Traverse over the kinds in freeVarsDict and `apply` them.
* Get all the `unknowns` for these kinds.
* Get `unknownsWithKinds` for these unknowns.

This is basically what you are doing, you are just manually calling substituteType rather than `apply`, and adding the additional `unknowns allTy`, which I'm not sure is necessary.

Reply:
I suppose maybe it's possible to have unknowns in allTy if a class was generalized with an unused/redundant type variable 😬.

Reply:
Yeah, I ripped it out and ran the test suite, which found some such cases. Here's one:

```purs
newtype Indexed ∷ forall k1 k2 k3. (k1 → Type) → k2 → k3 → k1 → Type
newtype Indexed m x y a = Indexed (m a)

class IxFunctor ∷ ∀ ix. (ix → ix → Type → Type) → Constraint
class IxFunctor f where
  imap ∷ ∀ a b x y. (a → b) → f x y a → f x y b

instance ixFunctorIndexed ∷ Functor m ⇒ IxFunctor (Indexed m) where
  imap f (Indexed ma) = Indexed (map f ma)
```

Here, the unknown shows up in the `KindApp`s in `allTy`, but not as the kind of any free variable.

Reply:
Is there a particular reason for the `IS.filter (`M.member` uns)` that makes it necessary?

Reply:
That might be a vestige of an earlier iteration. `unknownsWithKinds` (`lookupUnsolved`) crashes if it receives a solved unknown. Is the result of `substituteType`/`apply` guaranteed not to contain any? That would make sense, yeah.

Reply:
Yes, it recursively substitutes solved unknowns.